### PR TITLE
Add cross builtins and cross HLSL function to DirectX and SPIR-V backend

### LINKED ISSUE #99095: Implement the `cross` HLSL Function
- [ ] Implement `cross` clang builtin,
- [ ] Link `cross` clang builtin with `hlsl_intrinsics.h`
- [ ] Add sema checks for `cross` to `CheckHLSLBuiltinFunctionCall` in `SemaChecking.cpp`
- [ ] Add codegen for `cross` to `EmitHLSLBuiltinExpr` in `CGBuiltin.cpp`
- [ ] Add codegen tests to `clang/test/CodeGenHLSL/builtins/cross.hlsl`
- [ ] Add sema tests to `clang/test/SemaHLSL/BuiltIns/cross-errors.hlsl`
- [ ] Create the `int_spv_cross` intrinsic in `IntrinsicsSPIRV.td`
- [ ] In SPIRVInstructionSelector.cpp create the `cross` lowering and map  it to `int_spv_cross` in `SPIRVInstructionSelector::selectIntrinsic`.
- [ ] Create SPIR-V backend test case in `llvm/test/CodeGen/SPIRV/hlsl-intrinsics/cross.ll`

## DirectX

There were no DXIL opcodes found for `cross`.

## SPIR-V

# Cross:

## Description:
**Cross**  
  
Result is the cross product of *x* and *y*, i.e., the resulting
components are, in order:  

*x*\[1\] \* *y*\[2\] - *y*\[1\] \* *x*\[2\]  

*x*\[2\] \* *y*\[0\] - *y*\[2\] \* *x*\[0\]  

*x*\[0\] \* *y*\[1\] - *y*\[0\] \* *x*\[1\]  
  
All the operands must be vectors of 3 components of a floating-point
type.  
  
*Result Type* and the type of all operands must be the same type.

<table>
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr>
<th>Number</th>
<th>Operand 1</th>
<th>Operand 2</th>
<th>Operand 3</th>
<th>Operand 4</th>
</tr>
</thead>
<tbody>
<tr>
<td class="tableblock halign-left valign-top"><p>68</p></td>
<td
class="tableblock halign-left valign-top"><p><em>&lt;id&gt;</em><br />
<em>x</em></p></td>
<td
class="tableblock halign-left valign-top"><p><em>&lt;id&gt;</em><br />
<em>y</em></p></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>

### Example 1
```hlsl
//dxc cross_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export float3 fn(float3 p1, float3 p2) {
    return cross(p1, p2);
}
```
## HLSL:

Returns the cross product of two floating-point, 3D vectors.

| *ret* cross(*x*, *y*) |
|-----------------------|

 

## Parameters

| Item                                                   | Description                                             |
|--------------------------------------------------------|---------------------------------------------------------|
| <span id="x"></span><span id="X"></span>*x*<br/> | \[in\] The first floating-point, 3D vector.<br/>  |
| <span id="y"></span><span id="Y"></span>*y*<br/> | \[in\] The second floating-point, 3D vector.<br/> |

 

## Return Value

The cross product of the *x* parameter and the *y* parameter.

## Type Description

| Name  | [**Template Type**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)                       | [**Component Type**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md) | Size |
|-------|-------------------------------------------------------------------------------------|----------------------------------------------------------------|------|
| *x*   | [**vector**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md) | [**float**](/windows/desktop/WinProg/windows-data-types)                        | 3    |
| *y*   | [**vector**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md) | [**float**](/windows/desktop/WinProg/windows-data-types)                        | 3    |
| *ret* | [**vector**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md) | [**float**](/windows/desktop/WinProg/windows-data-types)                        | 3    |

 

## Minimum Shader Model

This function is supported in the following shader models.

| Shader Model                                                                       | Supported             |
|------------------------------------------------------------------------------------|-----------------------|
| [Shader Model 2 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm2.md) and higher shader models | yes                   |
| [Shader Model 1 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm1.md)                          | vs\_1\_1 and ps\_1\_4 |

 

## See also

<dl> <dt>

[**Intrinsic Functions (DirectX HLSL)**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)
</dt> </dl>

### PR DESCRIPTION
**Pull request by bob80905**
This PR adds the step intrinsic and an HLSL function that uses it.
The SPIRV backend is also implemented.

Used https://github.com/llvm/llvm-project/pull/106471 as a reference.
Fixes https://github.com/llvm/llvm-project/issues/99095

### CODE REVIEW

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -18639,6 +18639,21 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
         IsUnsigned ? Intrinsic::dx_uclamp : Intrinsic::dx_clamp,
         ArrayRef<Value *>{OpX, OpMin, OpMax}, nullptr, "dx.clamp");
   }
+  case Builtin::BI__builtin_hlsl_cross: {
+    Value *Op0 = EmitScalarExpr(E->getArg(0));
+    Value *Op1 = EmitScalarExpr(E->getArg(1));
+    assert(E->getArg(0)->getType()->hasFloatingRepresentation() &&
+           E->getArg(1)->getType()->hasFloatingRepresentation() &&
+           "step operands must have a float representation");
```
Comment (farzonl) [inline clang/lib/CodeGen/CGBuiltin.cpp:18647]:
```suggestion
           "cross operands must have a float representation");
```

File: llvm/include/llvm/IR/IntrinsicsDirectX.td [hunk 2]
```diff
@@ -44,6 +44,7 @@ def int_dx_cast_handle : Intrinsic<[llvm_any_ty], [llvm_any_ty]>;
 def int_dx_all : DefaultAttrsIntrinsic<[llvm_i1_ty], [llvm_any_ty], [IntrNoMem]>;
 def int_dx_any : DefaultAttrsIntrinsic<[llvm_i1_ty], [llvm_any_ty], [IntrNoMem]>;
 def int_dx_clamp : DefaultAttrsIntrinsic<[llvm_any_ty], [LLVMMatchType<0>, LLVMMatchType<0>, LLVMMatchType<0>], [IntrNoMem]>;
+def int_dx_cross : DefaultAttrsIntrinsic<[llvm_anyfloat_ty], [LLVMMatchType<0>, LLVMMatchType<0>], [IntrNoMem]>;
```
Comment (farzonl) [inline llvm/include/llvm/IR/IntrinsicsDirectX.td:47]: I think you are trying to keep this alphabetica, but  don't put this between the clamp intrinsics.

File: clang/lib/Sema/SemaHLSL.cpp [hunk 3]
```diff
@@ -1704,6 +1704,20 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
...
     break;
   }
+  case Builtin::BI__builtin_hlsl_cross: {
+    if (SemaRef.checkArgCount(TheCall, 2))
+      return true;
+    if (CheckVectorElementCallArgs(&SemaRef, TheCall))
+      return true;
+    if (CheckFloatOrHalfRepresentations(&SemaRef, TheCall))
+      return true;
+
```
Comment (farzonl) [inline clang/lib/Sema/SemaHLSL.cpp:1714]: Maybe we should do a  vector size check to prevent misuse of the builtin? Downsize is the check isn't triggerable from the hlsl function name.

File: llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp [hunk 4]
```diff
@@ -73,6 +74,42 @@ static Value *expandAbs(CallInst *Orig) {
...
+  VectorType *VT = cast<VectorType>(Orig->getType());
+  if (cast<FixedVectorType>(VT)->getNumElements() != 3)
+    report_fatal_error(Twine("return vector must have exactly 3 elements"),
+                       /* gen_crash_diag=*/false);
+
+  Value *op0 = Orig->getOperand(0);
+  Value *op1 = Orig->getOperand(1);
+  IRBuilder<> Builder(Orig);
+
+  Value *op0_x = Builder.CreateExtractElement(op0, (uint64_t)0);
```
Comment (farzonl) [inline llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:88]: NIT: in general I don't think this matters but since cross product is crossing over elements
with `x1*y2 - x2*y1` I think its worth naming these elements like so  `Builder.CreateExtractElement(op0, 0, "x0");`
This will effect the SSA names so it isn't just %1-%n

File: clang/lib/Sema/SemaHLSL.cpp (hunk 3 above)
Comment (bob80905) [reply clang/lib/Sema/SemaHLSL.cpp:1714]: There are also other asserts to catch when the size isn't 3, I don't think this is needed.

Comment (farzonl) [reply clang/lib/Sema/SemaHLSL.cpp:1714]: If thats the case could you add a test to `clang/test/CodeGenHLSL/builtins/cross.hlsl` As long as this gets caugh by an error then we should be good.

```hlsl
float2 builtin_cross_float2(float2  p1, float2  p2)
{
  return __builtin_hlsl_cross(p1, p2);
}
```

File: clang/test/SemaHLSL/BuiltIns/cross-errors.hlsl [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+bool builtin_cross_int_to_float_promotion(int p1)
+{
+  return __builtin_hlsl_cross(p1, p1);
+  // expected-error@-1 {{passing 'int' to parameter of incompatible type 'float'}}
+}
+
+bool2 builtin_cross_int2_to_float2_promotion(int2 p1)
+{
+  return __builtin_hlsl_cross(p1, p1);
+  // expected-error@-1 {{passing 'int2' (aka 'vector<int, 2>') to parameter of incompatible type '__attribute__((__vector_size__(2 * sizeof(float)))) float' (vector of 2 'float' values)}}
```
Comment (farzonl) [inline clang/test/SemaHLSL/BuiltIns/cross-errors.hlsl:30]: what happens if its 
```hlsl
float3  builtin_cross_float3_int3(float3 p1, int3 p2)
{
  return __builtin_hlsl_cross(p1, p2);
}
```

File: clang/lib/Sema/SemaHLSL.cpp (hunk 3 above)
Comment (bob80905) [reply clang/lib/Sema/SemaHLSL.cpp:1714]: I added a diagnostic and put the test in cross-errors.hlsl instead.

File: clang/test/SemaHLSL/BuiltIns/cross-errors.hlsl (hunk 5 above)
Comment (farzonl) [reply clang/test/SemaHLSL/BuiltIns/cross-errors.hlsl:30]: ah ok it got promoted in DXC so wanted t check: https://hlsl.godbolt.org/z/3nvGve7WP

File: llvm/test/CodeGen/DirectX/cross.ll [hunk 6]
```diff
@@ -0,0 +1,57 @@
+; RUN: opt -S  -dxil-intrinsic-expansion  < %s | FileCheck %s --check-prefix=CHECK
```
Comment (farzonl) [inline llvm/test/CodeGen/DirectX/cross.ll:1]: typicall we do `--check-prefix` if one run is different than the other.  in this case they are the same. My recommendation is to drop one of these runs  and remove the check prefix.

**Review by farzonl — APPROVED** [review]

File: llvm/test/CodeGen/DirectX/cross.ll (hunk 6 above)
Comment (farzonl) [reply llvm/test/CodeGen/DirectX/cross.ll]: My point still stands why do you need two runs?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 7]
```diff
@@ -312,6 +312,8 @@ def err_invalid_vector_long_double_decl_spec : Error<
   "cannot use 'long double' with '__vector'">;
 def err_invalid_vector_complex_decl_spec : Error<
   "cannot use '_Complex' with '__vector'">;
+def err_invalid_vector_size : Error<
+  "expected vector size of '%0', but vector size is '%1'">;
```
Comment (llvm-beanz) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:316]: I'm a little surprised we don't have a similar diagnostic, but it seems that we don't. Maybe we could make a small tweak to an existing message and reuse the message. What if we changed `err_vector_incorrect_num_initializers` to:

```
def err_vector_incorrect_num_initializers : Error<
  "%select{too many|too few}0 elements in vector %select{initialization|operand}3 (expected %1 elements, have %2)">;
```
Then you could add a `0` argument to the one place it is called, and change your new diagnostic to:

```suggestion
def err_incorrect_vector_element_count : Error<
  err_vector_incorrect_num_initializers.Summary>;
```

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 8]
```diff
@@ -10506,7 +10506,9 @@ def err_second_argument_to_cwsc_not_pointer : Error<
   "second argument to __builtin_call_with_static_chain must be of pointer type">;
 
 def err_vector_incorrect_num_initializers : Error<
-  "%select{too many|too few}0 elements in vector initialization (expected %1 elements, have %2)">;
+  "%select{too many|too few}0 elements in vector %select{initialization|operand}3 (expected %1 elements, have %2)">;
+def err_incorrect_vector_element_count  : Error<
```
Comment (llvm-beanz) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:10510]: I think you updated all the uses of this to use the other diagnostic, which is fine, but then we can remove this.

Comment (llvm-beanz) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10510]: Also maybe rename `err_vector_incorrect_num_initializers` -> `err_vector_incorrect_num_elements`?

File: clang/lib/Sema/SemaHLSL.cpp [hunk 9]
```diff
@@ -1828,6 +1828,41 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
...
+    if (NumElementsArg1 != 3) {
+      int LessOrMore = NumElementsArg1 > 3 ? 1 : 0;
+      SemaRef.Diag(TheCall->getBeginLoc(),
+                   diag::err_vector_incorrect_num_initializers)
+          << LessOrMore << 3 << NumElementsArg1 << 1;
+      return true;
+    }
+    if (NumElementsArg2 != 3) {
+      int LessOrMore = NumElementsArg1 > 3 ? 1 : 0;
```
Comment (bogner) [inline clang/lib/Sema/SemaHLSL.cpp:1852]: Should this be `NumElementsArg2`?

File: clang/include/clang/Basic/Builtins.td [hunk 10]
```diff
@@ -4709,6 +4709,18 @@ def HLSLClamp : LangBuiltin<"HLSL_LANG"> {
   let Prototype = "void(...)";
 }
 
+def HLSLCreateHandle : LangBuiltin<"HLSL_LANG"> {
+  let Spellings = ["__builtin_hlsl_create_handle"];
+  let Attributes = [NoThrow, Const];
+  let Prototype = "void*(unsigned char)";
+}
```
Comment (llvm-beanz) [inline clang/include/clang/Basic/Builtins.td:4716]: This looks like an unrelated change sliding back in?

File: clang/lib/Sema/SemaInit.cpp [hunk 11]
```diff
@@ -1977,7 +1977,7 @@ void InitListChecker::CheckVectorType(const InitializedEntity &Entity,
     if (!VerifyOnly)
       SemaRef.Diag(IList->getBeginLoc(),
                    diag::err_vector_incorrect_num_initializers)
-          << (numEltsInit < maxElements) << maxElements << numEltsInit;
+          << (numEltsInit < maxElements) << maxElements << numEltsInit << 0;
```
Comment (llvm-beanz) [inline clang/lib/Sema/SemaInit.cpp:1980]: nit: please comment the meaning of the number

```suggestion
          << (numEltsInit < maxElements) << maxElements << numEltsInit << /*initialization*/ 0;
```

File: clang/lib/Sema/SemaHLSL.cpp [hunk 12]
```diff
@@ -1828,6 +1828,41 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
...
+
+      SemaRef.Diag(TheCall->getBeginLoc(),
+                   diag::err_vector_incorrect_num_initializers)
+          << LessOrMore << 3 << NumElementsArg2 << 1;
```
Comment (llvm-beanz) [inline clang/lib/Sema/SemaHLSL.cpp:1856]: nit:
```suggestion
          << LessOrMore << 3 << NumElementsArg2 << /*operand*/ 1;
```

File: clang/lib/Sema/SemaHLSL.cpp [hunk 13]
```diff
@@ -1828,6 +1828,41 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
...
+    int NumElementsArg1 =
+        TheCall->getArg(0)->getType()->getAs<VectorType>()->getNumElements();
+    int NumElementsArg2 =
+        TheCall->getArg(1)->getType()->getAs<VectorType>()->getNumElements();
+
+    if (NumElementsArg1 != 3) {
+      int LessOrMore = NumElementsArg1 > 3 ? 1 : 0;
+      SemaRef.Diag(TheCall->getBeginLoc(),
+                   diag::err_vector_incorrect_num_initializers)
+          << LessOrMore << 3 << NumElementsArg1 << 1;
```
Comment (llvm-beanz) [inline clang/lib/Sema/SemaHLSL.cpp:1848]: nit:
```suggestion
          << LessOrMore << 3 << NumElementsArg1 << /*operand*/ 1;
```

File: clang/test/CodeGenHLSL/builtins/cross.hlsl [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+// RUN: %clang_cc1 -finclude-default-header -x hlsl -triple \
+// RUN:   spirv-unknown-vulkan-compute %s -emit-llvm -disable-llvm-passes \
+// RUN:   -o - | FileCheck %s --check-prefixes=CHECK,NO_HALF \
+// RUN:   -DFNATTRS="spir_func noundef" -DTARGET=spv
+
+// NATIVE_HALF: define [[FNATTRS]] <3 x half> @
+// NATIVE_HALF: call <3 x half> @llvm.[[TARGET]].cross.v3f16(<3 x half>
+// NO_HALF: call <3 x float> @llvm.[[TARGET]].cross.v3f32(<3 x float>
+// NATIVE_HALF: ret <3 x half> %hlsl.cross
+// NO_HALF: ret <3 x float> %hlsl.cross
```
Comment (bogner) [inline clang/test/CodeGenHLSL/builtins/cross.hlsl:24]: It's probably clearer to put the `NATIVE_HALF` checks consecutively, followed by the `NO_HALF` checks - the way they're intertwined here is a bit hard to read. Also, it looks like the `define` check is missing for the `NO_HALF` case.

**Review by llvm-beanz — COMMENTED** [review]
A few last small things, but mostly looks good.

File: llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp [hunk 15]
```diff
@@ -74,6 +75,42 @@ static Value *expandAbs(CallInst *Orig) {
...
+  Value *op0_z = Builder.CreateExtractElement(op0, 2, "x2");
+
+  Value *op1_x = Builder.CreateExtractElement(op1, (uint64_t)0, "y0");
+  Value *op1_y = Builder.CreateExtractElement(op1, 1, "y1");
+  Value *op1_z = Builder.CreateExtractElement(op1, 2, "y2");
+
+  auto MulSub = [&](Value *x0, Value *y0, Value *x1, Value *y1) -> Value * {
+    Value *xy = Builder.CreateFMul(x0, y1);
+    Value *yx = Builder.CreateFMul(y0, x1);
+    return Builder.CreateFSub(xy, yx);
```
Comment (bogner) [inline llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:100]: Should we copy the value name from `Orig` for the final result here?

File: clang/include/clang/Basic/Builtins.td (hunk 10 above)
Comment (bob80905) [reply clang/include/clang/Basic/Builtins.td:4716]: It's the result of a merge, I'll remove it.

File: clang/lib/Sema/SemaHLSL.cpp (hunk 9 above)
Comment (bob80905) [reply clang/lib/Sema/SemaHLSL.cpp:1852]: Good catch!

File: llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp (hunk 15 above)
Comment (bob80905) [reply llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:100]: I don't think so, since we want distinct variable names when constructing the final result vector. Each insertelement needs a unique variable name so that contents of the vector are correct.

**Review by damyanp — APPROVED** [review]
LGTM, and I see you've addressed all of the feedback.

Would be good to check @bogner is happy with the resolution of [this issue](https://github.com/llvm/llvm-project/pull/109180#discussion_r1781967896) before merging.

File: clang/test/SemaHLSL/BuiltIns/cross-errors.hlsl [hunk 16]
```diff
@@ -0,0 +1,43 @@
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -disable-llvm-passes -verify -verify-ignore-unexpected
```
Comment (pow2clk) [inline clang/test/SemaHLSL/BuiltIns/cross-errors.hlsl:1]: Is this producing or expected to produce unexpected diagnostics? I see this flag in a few other uses of this in SemaHLSL, but not so much in other Sema* test directories. Perhaps it's getting copied over needlessly? 

If there are some tricky diagnostics getting produced, it might be better to limit them by assigning tis to a specific type of diagnostic.

File: llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp (hunk 15 above)
Comment (bob80905) [reply llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp]: I tried it out and turns out each variable was assigned a unique variable name, contrary to what I expected. So your comment has been incorporated.

File: clang/test/CodeGenHLSL/builtins/cross.hlsl [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+// RUN:   -DFNATTRS=noundef -DTARGET=dx
+// RUN: %clang_cc1 -finclude-default-header -x hlsl -triple \
+// RUN:   spirv-unknown-vulkan-compute %s -fnative-half-type \
+// RUN:   -emit-llvm -disable-llvm-passes -o - | FileCheck %s \
+// RUN:   --check-prefixes=CHECK,NATIVE_HALF \
+// RUN:   -DFNATTRS="spir_func noundef" -DTARGET=spv
+// RUN: %clang_cc1 -finclude-default-header -x hlsl -triple \
+// RUN:   spirv-unknown-vulkan-compute %s -emit-llvm -disable-llvm-passes \
+// RUN:   -o - | FileCheck %s --check-prefixes=CHECK,NO_HALF \
+// RUN:   -DFNATTRS="spir_func noundef" -DTARGET=spv
```
Comment (pow2clk) [inline clang/test/CodeGenHLSL/builtins/cross.hlsl:18]: It's harmless, but given that these run lines are long and hard to compare as-is, I'll note for next time that `-x hlsl` isn't necessary as long as the file ends in `.hlsl`

File: llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp [hunk 18]
```diff
@@ -74,6 +75,42 @@ static Value *expandAbs(CallInst *Orig) {
...
+  VectorType *VT = cast<VectorType>(Orig->getType());
+  if (cast<FixedVectorType>(VT)->getNumElements() != 3)
+    report_fatal_error(Twine("return vector must have exactly 3 elements"),
+                       /* gen_crash_diag=*/false);
+
+  Value *op0 = Orig->getOperand(0);
+  Value *op1 = Orig->getOperand(1);
+  IRBuilder<> Builder(Orig);
+
+  Value *op0_x = Builder.CreateExtractElement(op0, (uint64_t)0, "x0");
```
Comment (pow2clk) [inline llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:99]: Just curious why the 0 needs a cast and the others don't.

**Review by pow2clk — APPROVED** [review]
Looks good to me. I just have a couple nits that you might include if you create another commit for another reason.

File: clang/test/SemaHLSL/BuiltIns/cross-errors.hlsl (hunk 16 above)
Comment (bob80905) [reply clang/test/SemaHLSL/BuiltIns/cross-errors.hlsl:1]: Thanks, this revealed a missing open curly brace on line 18!
I got rid of this extra flag, seems to be a needless copy.

File: llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp (hunk 18 above)
Comment (bob80905) [reply llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:99]: Without the cast I get:
`error C2668: 'llvm::IRBuilderBase::CreateExtractElement': ambiguous call to overloaded function note: while trying to match the argument list '(llvm::Value *, int)' `
as a compilation error.

Comment (damyanp) [reply llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:99]: Isn't C++ fun!

**Review by bogner — APPROVED** [review]
